core/lib/dal/src/basic_witness_input_producer_dal.rs

```diff
@@ -0,0 +1,168 @@
+use crate::instrument::InstrumentExt;
+use crate::time_utils::{duration_to_naive_time, pg_interval_from_duration};
+use crate::StorageProcessor;
+use sqlx::postgres::types::PgInterval;
+use std::time::{Duration, Instant};
+use zksync_types::L1BatchNumber;
+
+#[derive(Debug)]
+pub struct BasicWitnessInputProducerDal<'a, 'c> {
+    pub(crate) storage: &'a mut StorageProcessor<'c>,
+}
+
+/// The amount of attempts to process a job before giving up.
+const JOB_MAX_ATTEMPT: i16 = 10;
+
+/// Time to wait for job to be processed
+const JOB_PROCESSING_TIMEOUT: PgInterval = pg_interval_from_duration(Duration::from_secs(10 * 60));
+
+/// Status of a job that the producer will work on.
+#[derive(Debug, strum::Display, strum::EnumString, strum::AsRefStr)]
+pub enum BasicWitnessInputProducerJobStatus {
+    /// When the job is queued. Metadata calculator creates the job and marks it as queued.
+    #[strum(serialize = "queued")]
+    Queued,
+    /// The job is not going to be processed. This state is designed for manual operations on DB.
+    /// It is expected to be used if some jobs should be skipped like:
+    /// - testing purposes (want to check a specific L1 Batch, I can mark everything before it skipped)
+    /// - trim down costs on some environments (if I've done breaking changes,
+    /// makes no sense to wait for everything to be processed, I can just skip them and save resources)
+    #[strum(serialize = "manually_skipped")]
+    ManuallySkipped,
+    /// Currently being processed by one of the jobs. Transitory state, will transition to either
+    /// [`BasicWitnessInputProducerStatus::Successful`] or [`BasicWitnessInputProducerStatus::Failed`].
+    #[strum(serialize = "in_progress")]
+    InProgress,
+    /// The final (happy case) state we expect all jobs to end up. After the run is complete,
+    /// the job uploaded it's inputs, it lands in successful.
+    #[strum(serialize = "successful")]
+    Successful,
+    /// The job failed for reasons. It will be marked as such and the error persisted in DB.
+    /// If it failed less than MAX_ATTEMPTs, the job will be retried,
+    /// otherwise it will stay in this state as final state.
+    #[strum(serialize = "failed")]
+    Failed,
+}
+
+impl BasicWitnessInputProducerDal<'_, '_> {
+    pub async fn create_basic_witness_input_producer_job(
+        &mut self,
+        l1_batch_number: L1BatchNumber,
+    ) -> anyhow::Result<()> {
```

Comment:
Is using `anyhow::Result` as opposed to `Result<_, sqlx::Error>` a conscious decision here? AFAICT, the existing DALs use `anyhow::Result` only if errors are generated in DAL logic. (I'd argue that some of these cases could be replaced with [`sqlx::Error::Decode`](https://docs.rs/sqlx/latest/sqlx/enum.Error.html#variant.Decode) / `ColumnDecode`.